srk/src/numberField.mli

```diff
@@ -0,0 +1,214 @@
+open Polynomial
+
+(** Given a univariate polynomial, and a variable dimension, create an
+    equivalent multivariate polynomial using the given dimension.*)
+val make_multivariate : Monomial.dim -> QQX.t -> QQXs.t
+
+(** Given a multivariate polynomial over a single dimension, create an
+    equivalent univariate polynomial.
+    @raise Failure if input is not a univariate polynomial.*)
+val make_univariate : QQXs.t -> QQX.t
+
+(** Let K1 = Q[v0,v1]/p and K2=Q[v0,v1]/q be number fields. Let K1 have degree d1 and K2 have degree d2.
+    [prim, v0_in_prim, v1_in_prim] = [primitive_elem (d1*d2) p q v0 v1] 
+    is such that the number field K3 = Q[x]/prim of degree d1*d2 such that K3 contains K1 and K2. 
+    [v0_in_prim] and [v1_in_prim] give the representations of v0 and v1 in K3 respectively. For
+    the method to work p needs to be irreducible in K2 and q needs to be irreducible in K1.*)
+val primitive_elem : int -> QQXs.t -> QQXs.t -> Monomial.dim -> Monomial.dim -> QQX.t * QQX.t * QQX.t
```

Comment:
This is messy, and I couldn't quite come up with a good interface. If all you want is to compute a primitive element for two number fields Q[x]/p and Q[y]/q, then variables and such don't need to be stated. However, for factoring and computing a splitting field we want to compute a "primitive element" for something like Q[x,y]/<p(x),q(x, y)>. In this case the y roots of q(x,y) depend on x. So we need both variables. You probably could simplify the interface to only computing a primitive element for fields of the form Q[x]/p(x). But this would be extra expensive for the cases that are needed for factoring and computing a splitting field.

Reply:
So should I think of Q[x]/p as a representation of  the number field Q(a) for some a, and Q[x,y]/<p(x),q(x,y)> as a representation of Q(a,b) where b is a root of q(x,y), considered as a univariate polynomial over Q(a)?

Reply:
Yes. This function is a little messy. We could keep it internal to the .ml file if you think the interface is too complicated.

Reply:
I think it's ok as an internal function.
I'm still not quite sure that I understand what it's doing: we have two number fields
Q(a,b) and Q(c,d), represented as Q[v0,v1]/<p(v0),q(v0, v1)> and Q[v0,v1]/<r(v0),s(v0, v1)>.  We want a primitive element e such that Q(e) contains both Q(a,b) and Q(c,d).
- Is d1 the degree of Q(a,b) as a extension of Q(a) or Q?
- We only take q and s as input -- p and r are not needed?  (Do we require p = r?)
- What are v0_in_prim / v1_in_prim?

Reply:
Examples
 - Consider K_1 = Q(sqrt(2)) represented as Q[x]/<x^2-2> and K_2 = Q(sqrt(3)) represented as Q[y]/<y^2 - 3>. A field that contains both is K_3 = Q(sqrt(2), sqrt(3)) which can be represented by Q[x, y]/<x^2-2, y^2-3>. However, K_3 is not represented by a primitive element. The goal of the function is to find a p(z) such that K_3 = Q[z]/<p(z)>. For this example needed_deg = 4, mp0 = x^2 - 2, mp1 = y^2 - 3, v0 = x, and v1 = y. The result is p(z) = z^4 - 10z^2 + 1, which has deg, 4. v0_in_prim is a representation of x in the new field K3. For this case v0_in_prim = (z^3 - 9z)/2 and v1_in_prim = -(z^3-11z)/2. That is (v0_in_prim^2 - 2) mod <z^4 - 10z^2 + 1> = 0 and (v1_in_prim^2 - 3) mod <z^4 - 10z^2 + 1> = 0. 
 - Consider K_1 = Q[x]/<x^3-2>. Consider q(x, y) = y^2+xy+x^2 as an irreducible polynomial of K_1[y], and thus doesn't factor over K_1. We wish to enlarge K_1 with a root of q(x, y). Such a field is K3 = Q[x, y]/<x^3-2, y^2+xy+x^2>. We wish to find a primitive element that represents K3. In this case needed_deg = 6, mp0 = x^3-2, mp1 = y^2+xy+x^2, v0 = x, v1 = y. The result is p(z) = z^6 + 108. I.e. K3 = Q[z]/<z^6 + 108>. Let v0_in_prim = a(z) and v1_in_prim = b(z). Then (a(z)^3 - 2) mod <z^6 + 108> = 0 and (b(z)^2 + a(z)b(z) + a(z)^2) mod <z^6 + 108> = 0.
 - Consider K = Q[x]/<x^6 + 108>. We wish to factor y^3 - 2 in K. Consider L = K[y]/<y^3-2> = Q[x,y]<x^6 + 108, y^3-2>. L is *not* a field, because y^3 - 2 is reducible in K. Nevertheless, if we call primitive_elem 18 (x^6 + 108) (y^3 - 2) x y we will obtain a representation of L as Q[z]/p(z). As it turns out p(z) will factor over Q as three factors. Within each factor we can use v0_in_prim and v1_in_prim to get a polynomial f(x, y), which will be a factor of y^3 - 2 in K.
 
These are the three uses of primitive_elem. The first example is for "joining" fields. The second example, extends a field with an additional root. This is used in computing a splitting field. The final example doesn't compute a field at all, but is used when factoring over a number field. This is what's messy about the interface for this function. Conceptually, it is used for different things. However, all three examples perform the same computational steps.

Reply:
Perhaps this makes more sense in the language of commutative algebra.  Given an irreducible polynomial p in Q[v0] and a polynomial q in Q[v0,v1] \ Q[v0], we wish to find a polynomial r in Q[x] such that Q[x]/r is isomorphic to Q[v0,v1]/<p,q>.  The isomorphism is represented as v0_in_prim and v1_in_prim, which are the images of the two generators v0 + <p,q> and v_1 + <p,q> of Q[v0,v1]/<p,q> under the isomorphism.  Is that an accurate description?

Reply:
Yes, that seems right.

Reply:
Changed documentation.